src/mpid/common/genq/mpidu_genq_shmem_queue.h

```diff
@@ -83,103 +155,92 @@ static inline MPIDU_genqi_shmem_cell_header_s
-            *cell = HEADER_TO_CELL(cell_h);
-            queue->q.head.s = cell_h->next;
-        }
+    } else {
+        cell_h = HANDLE_TO_HEADER(base, queue->q.head.s);
+        *cell = HEADER_TO_CELL(cell_h);
+        queue->q.head.s = cell_h->u.inverse_queue.next;
     }
 
-  fn_exit:
-    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_GENQ_SHMEM_QUEUE_DEQUEUE);
     return rc;
 }
 
-static inline int MPIDU_genq_shmem_queue_enqueue(MPIDU_genq_shmem_queue_t queue, void *cell)
+static inline int genqi_inv_mpsc_enqueue(void *base, genqi_shmem_queue_u * queue, void *cell)
 {
     int rc = MPI_SUCCESS;
-    MPIDU_genqi_shmem_cell_header_s *cell_h = CELL_TO_HEADER(cell);
+    genqi_shmem_cell_header_s *cell_h = CELL_TO_HEADER(cell);
+    cell_h->u.inverse_queue.next = 0;
+    cell_h->u.inverse_queue.prev = 0;
 
-    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_GENQ_SHMEM_QUEUE_ENQUEUE);
-    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDU_GENQ_SHMEM_QUEUE_ENQUEUE);
+    void *prev_handle = NULL;
+    void *handle = (void *) cell_h->handle;
 
-    if (queue->q.flags == MPIDU_GENQ_SHMEM_QUEUE_TYPE__SERIAL) {
-        uintptr_t handle = cell_h->handle;
+    do {
+        prev_handle = MPL_atomic_load_ptr(&queue->q.tail.m);
+        cell_h->u.inverse_queue.prev = (uintptr_t) prev_handle;
+    } while (MPL_atomic_cas_ptr(&queue->q.tail.m, prev_handle, handle) != prev_handle);
 
-        if (queue->q.tail.s) {
-            HANDLE_TO_HEADER(queue->q.pool, queue->q.tail.s)->next = handle;
-        }
-        cell_h->prev = queue->q.tail.s;
-        queue->q.tail.s = handle;
-        if (!queue->q.head.s) {
-            queue->q.head.s = queue->q.tail.s;
-        }
-    } else {    /* MPIDU_GENQ_SHMEM_QUEUE_TYPE__MPSC */
-        void *prev_handle = NULL;
-        void *handle = (void *) cell_h->handle;
-
-        do {
-            prev_handle = MPL_atomic_load_ptr(&queue->q.tail.m);
-            cell_h->prev = (uintptr_t) prev_handle;
-        } while (MPL_atomic_cas_ptr(&queue->q.tail.m, prev_handle, handle) != prev_handle);
-    }
-
-    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_GENQ_SHMEM_QUEUE_ENQUEUE);
     return rc;
 }
 
-static inline void *MPIDU_genq_shmem_queue_head(MPIDU_genq_shmem_queue_t queue)
-{
-    MPIDU_genqi_shmem_cell_header_s *cell_h = NULL;
+/* EXTERNAL INTERFACE */
 
-    if (queue->q.flags == MPIDU_GENQ_SHMEM_QUEUE_TYPE__SERIAL) {
-        cell_h = HANDLE_TO_HEADER(queue->q.pool, queue->q.head.s);
-    } else {    /* MPIDU_GENQ_SHMEM_QUEUE_TYPE__MPSC */
-        if (!queue->q.head.s) {
-            cell_h = MPIDU_genqi_shmem_get_head_cell_header(queue);
-        } else {
-            cell_h = HANDLE_TO_HEADER(queue->q.pool, queue->q.head.s);
-        }
-
-    }
-    if (cell_h) {
-        return HEADER_TO_CELL(cell_h);
+static inline int MPIDU_genq_shmem_queue_dequeue(MPIDU_genq_shmem_queues_t queues,
+                                                 int queue_idx, void **cell)
```

Comment:
This interface change is strange. It is not a general queue interface. It seems have a specific usecase in mind.

Reply:
This queues + idx interface is quite counter-intuitive to general queue.

Reply:
Otherwise, you need expose some internal details. Shmem queue is unique from normal queues, I think one can get used to it.